torch/_cuda_sanitizer/_sanitizer.py

```diff
@@ -0,0 +1,490 @@
+import enum
+import functools
+import io
+import logging
+import sys
+import textwrap
+import traceback
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterator, List, Optional, Set, Tuple, TypeVar
+
+import torch
+import torch.utils._cuda_trace as cuda_trace
+from torch.utils._cuda_trace import _CUDA_HOOK_TRACE_OFFSET
+from torch.utils._python_dispatch import (
+    TorchDispatchMode,
+    _PYTHON_DISPATCH_TRACE_OFFSET,
+)
+from torch.utils._pytree import tree_map
+
+
+TK = TypeVar("TK")
+TVa = TypeVar("TVa")
+TVb = TypeVar("TVb")
+
+DataPtr = int
+StreamId = int
+EventId = int
+SeqNum = int
+
+logger = logging.getLogger(__name__)
+
+
+class AccessType(enum.Enum):
+    READ = enum.auto()
+    WRITE = enum.auto()
+
+    def __str__(self):
+        return "reading from" if self is AccessType.READ else "writing to"
+
+
+@dataclass
+class Access:
+    type: AccessType
+    seq_num: SeqNum
+    stream: StreamId
+    operator: str
+    names: List[str]
+    stack_trace: List[str]
```

Comment:
@zdevito you had advice for capturing Python stack traces efficiently, right?

Reply:
@zdevito  Thank you for the suggestion!

I also wanted to ask if you knew of a way to retrieve the line from the source code, given only the bytecode + lasti pair. I know that the line number of the trace can be easily retrieved, but wasn't able to find a way of capturing the source code.

@ezyang I thought it might improve readability to also print the source line in the error message, but perhaps the efficiency is more important?

Reply:
I think the code should retain the info; the point is to not spend a lot of time uselessly formatting stack traces when they never get seen by the user.

Reply:
I've ended up using the `traceback.StackSummary.extract` method. I looked at its source code, and it performs only the necessary operations for it to be formatted later.